### PR TITLE
 #20488 core dictionary get key

### PR DESCRIPTION
**Pull request by malcolmhoward**
This is an improved version of the original pull request (https://github.com/godotengine/godot/pull/20592) for issue https://github.com/godotengine/godot/issues/20488, which correctly targets the master branch as the base branch.

### CODE REVIEW

**Comment by aaronfranke** [comment]
Please read up on [pull request workflow](https://godot.readthedocs.io/en/latest/community/contributing/pr_workflow.html), specifically, [rebasing](https://godot.readthedocs.io/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase).

You can use `git rebase -i HEAD~4` and change `pick` to `fixup` to squash your commits.

**Comment by williamd1k0** [comment]
You just ignored my comment about `default` param lol

**Comment by malcolmhoward** [comment]
@aaronfranke @vnen Thank you for the suggestions.  I think this may be good to go now.

**Comment by malcolmhoward** [comment]
@williamd1k0  Whoops.  I'm sorry about that.  I just saw your comment.  The initial recommendation was to call get_valid from get, so I went with that for the sake of simplicity.  Thanks for the suggestions!

**Comment by vnen** [comment]
> The initial recommendation was to call get_valid from get, so I went with that for the sake of simplicity.

Just to be clear, I only said that because you said you copy-pasted the code. If you're changing the behavior then it's okay to make a copy to change the small bits.

**Comment by Faless** [comment]
> You just ignored my comment about default param lol

Agreed, a `default=null` param would be nice, and in line with what we are doing in other places:
http://docs.godotengine.org/en/3.0/classes/class_configfile.html#class-configfile-get-value

**Comment by williamd1k0** [comment]
@malcolmhoward Thanks, it looks like you're *new* to free software contributions. Remember to take discussions into account beforehand.

@Faless Thanks for pointing out an example for the `default` param.

**Comment by williamd1k0** [comment]
>The getptr() solution returned Null whenever the default argument was not passed, regardless of whether the key was valid.

I don't understand what you mean. I've done some assertions and it seems to work correctly.

```gdscript
var dict = {
	'a': 10,
	'c': false,
}

func _ready():
	assert(dict.get('a') == 10)
	assert(dict.get('b', 456) == 456)
	assert(dict.get('c', 789) == false)
	assert(dict.get('d') == null)
	assert(dict.get(Vector2(), true) == true)
```

**Comment by williamd1k0** [comment]
Btw, your approach is a bit slower than mine. Probably because you are searching for the key/value 2 times (https://github.com/godotengine/godot/pull/20627/files#diff-85daaf42c2e9eb94d22a705746ca7c32R117).

**Comment by malcolmhoward** [comment]
`var dict_1 = {"name": "John", "age": 22, "foo": "bar"}
	print("Name: ", dict_1["name"],
	" Age: ", dict_1["age"],
	" foo1: ", dict_1.get("foo"),
	" foo2: ", dict_1.get("foo", "test1"),
	" foo3: ", dict_1.get("foo", 'test2'),
	" foo4: ", dict_1.get("foot", 'test3'),
	" foo5: ", dict_1.get("foot")
	)
	# Name: John Age: 22 foo1: True foo2: True foo3: True foo4: test3 foo5: Null
	# ^^^^^ Output when Dictionary.get() uses getptr()
	# Name: John Age: 22 foo1: bar foo2: bar foo3: bar foo4: test3 foo5: Null
	# ^^^^^ Output when Dictionary.get() uses get_valid()

	var dict_2 = {
		'a': 10,
		'c': false,
	}
	
	assert(dict_2.get('a') == true)
	assert(dict_2.get('a') == 10)
	# ^^^^^ Error: Invalid operands 'bool' and 'int' in operator '==' when
	# dict.get() uses getptr() instead of get_valid()
	assert(dict_2.get('b', 456) == 456)
	assert(dict_2.get('c', 789) == false)
	assert(dict_2.get('d') == null)
	assert(dict_2.get(Vector2(), true) == true)`

Here's an example script that uses the assertions.  I apologize for getting my details confused earlier.  When Dictionary.get() in my solution uses getptr() instead of get_valid(), it returns True if the key exists, Null if the key doesn't exist, and the optional default value if one is provided.  I've used getptr() in my pull request to clarify.  Is there some way I can modify my solution to make getptr() return the value for the key instead of True?

File: core/dictionary.cpp [hunk 1]
```diff
@@ -112,6 +112,23 @@ Variant Dictionary::get_valid(const Variant &p_key) const {
...
 
+
+// When using getptr(p_key) instead of get_valid(p_key), this method returns True, p_default, or Null instead of
+// dict[p_key], p_default, or Null.
+Variant Dictionary::get(const Variant &p_key, const Variant &p_default) const {
+	// Using *result throws the following compilation error:
+	// invalid conversion from 'const Variant*' to 'Variant*' [-fpermissive]
+	// The result could be cast to some other type, but would get_valid(p_key) be simpler?
+	// Variant *result = getptr(p_key);
+	Variant result = getptr(p_key);
```
Comment (williamd1k0) [inline core/dictionary.cpp]: Did you tried adding `const` here?`const Variant result = getptr(p_key);`?

File: core/dictionary.cpp [hunk 2]
```diff
@@ -112,6 +112,23 @@ Variant Dictionary::get_valid(const Variant &p_key) const {
 	return E.get();
 }
 
+
```
Comment (Faless) [inline core/dictionary.cpp]: You also have an extra newline here which causes the static checks to fails

File: core/dictionary.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (malcolmhoward) [reply core/dictionary.cpp]:
```
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x354b0) [0x7ff65cd8a4b0] (??:0)
[2] Variant::reference(Variant const&) (godot/core/variant.cpp:909)
[3] Variant::Variant(Variant const&) (godot/core/variant.cpp:2640)
[4] Dictionary::get(Variant const&, Variant const&) const (godot/core/dictionary.cpp:125)
[5] _VariantCall::_call_Dictionary_get(Variant&, Variant&, Variant const**) (godot/core/variant_call.cpp:486)
[6] Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) (godot/core/variant_call.cpp:119)
[7] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) (godot/modules/gdscript/gdscript_function.cpp:1053)
[8] GDScriptInstance::_ml_call_reversed(GDScript*, StringName const&, Variant const**, int) (godot/modules/gdscript/gdscript.cpp:1182)
[9] GDScriptInstance::call_multilevel_reversed(StringName const&, Variant const**, int) (godot/modules/gdscript/gdscript.cpp:1191)
[10] Node::_notification(int) (godot/scene/main/node.cpp:139)
[11] Node::_notificationv(int, bool) (godot/./scene/main/node.h:46 (discriminator 14))
[12] CanvasItem::_notificationv(int, bool) (godot/./scene/2d/canvas_item.h:140 (discriminator 1))
[13] Control::_notificationv(int, bool) (godot/./scene/gui/control.h:51 (discriminator 1))
[14] Panel::_notificationv(int, bool) (godot/./scene/gui/panel.h:40 (discriminator 1))
[15] Object::notification(int, bool) (godot/core/object.cpp:930)
[16] Node::_propagate_ready() (godot/core/variant.h:413)
[17] Node::_propagate_ready() (godot/scene/main/node.cpp:173)
[18] Node::_set_tree(SceneTree*) (godot/scene/main/node.cpp:2441)
[19] SceneTree::init() (godot/scene/main/scene_tree.cpp:457)
[20] OS_X11::run() (godot/platform/x11/os_x11.cpp:2633)
[21] godot/bin/godot.x11.tools.64(main+0xd3) [0xed1909] (godot/platform/x11/godot_x11.cpp:56)
[22] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x7ff65cd75830] (??:0)
[23] godot/bin/godot.x11.tools.64(_start+0x29) [0xed1769] (??:?)
-- END OF BACKTRACE --
```

I get the above backtrace when I do

`Variant result = *getptr(p_key);`

I'm still a little confused why get_ptr(p_key) is preferable to get_valid(p_key), especially when get_valid(p_key) seems to return the expected result.  Has anyone been able to replicate the getptr() bug that I've experienced using the code from my branch?

**Comment by Faless** [comment]
Mmhhh... maybe:
```
Variant *result = getptr(p_key);
if (!result)
...
return *result;
```

**Comment by williamd1k0** [comment]
@malcolmhoward Are you still working on it?

**Comment by malcolmhoward** [comment]
@williamd1k0. Yep, thanks for checking.  I'll get this finished this week.

File: core/variant_call.cpp [hunk 3]
```diff
@@ -1654,6 +1655,7 @@ void register_variant_methods() {
 	ADDFUNC0R(DICTIONARY, ARRAY, Dictionary, keys, varray());
 	ADDFUNC0R(DICTIONARY, ARRAY, Dictionary, values, varray());
 	ADDFUNC1R(DICTIONARY, DICTIONARY, Dictionary, duplicate, BOOL, "deep", varray(false));
+	ADDFUNC2R(DICTIONARY, NIL, Dictionary, get, NIL, "key", NIL, "default", varray(Variant()));
```
Comment (aaronfranke) [inline core/variant_call.cpp]: I'm fairly new to working with Variant, but isn't the `m_ret` argument set to `NIL` here when it shouldn't be? See [Line 1410](https://github.com/godotengine/godot/blob/c89bc047ef2f12e4acee4b025de74bb83a8fee62/core/variant_call.cpp#L1410).

Comment (williamd1k0) [reply core/variant_call.cpp]: I'm also new to working with Variant, but in this case I think the `NIL` is acting as an "ANY Variant".

Example:
https://github.com/godotengine/godot/blob/c89bc047ef2f12e4acee4b025de74bb83a8fee62/core/variant_call.cpp#L1671-L1672

File: core/dictionary.cpp [hunk 4]
```diff
@@ -112,6 +112,23 @@ Variant Dictionary::get_valid(const Variant &p_key) const {
...
 }
 
+Variant Dictionary::get(const Variant &p_key, const Variant &p_default) const {
+	// When using getptr(p_key) instead of get_valid(p_key), this method returns True, p_default, or Null instead of
+	// dict[p_key], p_default, or Null.
+	// Using *result throws the following compilation error:
+	// invalid conversion from 'const Variant*' to 'Variant*' [-fpermissive]
+	// The result could be cast to some other type, but would get_valid(p_key) be simpler?
+	// Variant *result = getptr(p_key);
+	 Variant result = *getptr(p_key);
```
[the remark this answers is not in the document]
Comment (malcolmhoward) [reply core/dictionary.cpp]: @LikeLakers2 Thanks for pointing that out.   I completely missed it, and I'm fixing it now.

**Comment by aaronfranke** [comment]
You have quite a lot of squashing ahead of you, @malcolmhoward

**Comment by malcolmhoward** [comment]
@aaronfranke Thanks for catching that.  I forgot to --force push after rebasing from upstream.  :)

**Comment by malcolmhoward** [comment]
Is anyone opposed to me using get_valid() instead of getptr(), since get_valid() returns the desired result?  I'm still having difficulty with troubleshooting why getptr() doesn't work, when the logic for both methods appear to be very similar (see below).  I feel like I'm missing something simple, but I'm not (yet) familiar enough with C++ syntax to spot the issue.  I have tried all of the suggestions so far, but they either result in a boolean value of true being returned instead of the valid value, or they crash the editor and game altogether in Linux Mint (possible related to https://github.com/godotengine/godot/issues/22790).  Any help getting this across the finish line would be greatly appreciated!:

```
Variant *Dictionary::getptr(const Variant &p_key) {

	OrderedHashMap<Variant, Variant, VariantHasher, VariantComparator>::Element E = _p->variant_map.find(p_key);

	if (!E)
		return NULL;
	return &E.get();
}

Variant Dictionary::get_valid(const Variant &p_key) const {

	OrderedHashMap<Variant, Variant, VariantHasher, VariantComparator>::ConstElement E = ((const OrderedHashMap<Variant, Variant, VariantHasher, VariantComparator> *)&_p->variant_map)->find(p_key);

	if (!E)
		return Variant();
	return E.get();
}

Variant Dictionary::get(const Variant &p_key, const Variant &p_default) const {
	// When using getptr(p_key) instead of get_valid(p_key), this method returns True, p_default, or Null instead of
	// dict[p_key], p_default, or Null.
	// Using *result throws the following compilation error:
	// invalid conversion from 'const Variant*' to 'Variant*' [-fpermissive]
	// The result could be cast to some other type, but would get_valid(p_key) be simpler?
	// Variant *result = getptr(p_key);
	 Variant result = *getptr(p_key);
	// Variant result = getptr(p_key);
	// Variant result = get_valid(p_key);
	if (!result) {
		return p_default;
	}
	// return *result;
	return result;
}
```

**Comment by bojidar-bg** [comment]
Use `const Variant *result` instead of `Variant *result`?

**Comment by Faless** [comment]
Well, this should do it.
```
Variant Dictionary::get(const Variant &p_key, const Variant &p_default) const {
	const Variant *result = getptr(p_key);
	if (!result) {
		return p_default;
	}
	return *result;
}
```

**Comment by malcolmhoward** [comment]
> Well, this should do it.
> 
> ```
> Variant Dictionary::get(const Variant &p_key, const Variant &p_default) const {
> 	const Variant *result = getptr(p_key);
> 	if (!result) {
> 		return p_default;
> 	}
> 	return *result;
> }
> ```

@Faless That worked.  Thank you!  Now I just need to squash the remaining commits to get this down to one.

File: core/dictionary.cpp [hunk 5]
```diff
@@ -112,6 +112,15 @@ Variant Dictionary::get_valid(const Variant &p_key) const {
 	return E.get();
 }
 
+Variant Dictionary::get(const Variant &p_key, const Variant &p_default) const {
+	const Variant *result = getptr(p_key);
+	if (!result) {
+		return p_default;
+	}
+
+ 	return *result;
```
Comment (LikeLakers2) [inline core/dictionary.cpp]: You have a mix of tabs and spaces on this line. https://travis-ci.org/godotengine/godot/jobs/440229249#L536

Comment (malcolmhoward) [reply core/dictionary.cpp]: Thank you for finding that.  I just fixed it.

File: doc/classes/Dictionary.xml [hunk 6]
```diff
@@ -87,6 +87,14 @@
 				Return the list of values in the [code]Dictionary[/code].
 			</description>
 		</method>
+		<method name="get">
+			<return type="Variant">
+			</return>
+			<description>
```
Comment (akien-mga) [inline doc/classes/Dictionary.xml]: The argument seems to be missing:
```
<argument index="0" name="default" type="Variant">
</argument>
```

**Comment by Willowblade** [comment]
Damn, seems I should be more thorough in my search next time and not dismiss after two months in the search. But now it gets added and I'm a happy man! Thanks Rémi.

File: doc/classes/Dictionary.xml [hunk 7]
```diff
@@ -87,6 +87,14 @@
...
+				Returns the current value for the specified key in the [code]Dictionary[/code]. If the key does not
+				exist, the method returns the value of the optional default argument, or Null if it is omitted.
```
[the remark this answers is not in the document]
Comment (malcolmhoward) [reply doc/classes/Dictionary.xml]: @akien-mga Thanks for the suggestion.  I just updated the PR to remove the line break.

File: doc/classes/Dictionary.xml (hunk 6 above)
Comment (malcolmhoward) [reply doc/classes/Dictionary.xml]: @akien-mga Thanks for the suggestion.  I just updated the PR to add the missing arguments in the .get() method documentation.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**